jaxopt/_src/linear_solve.py

```diff
@@ -141,16 +153,34 @@ def solve_normal_cg(matvec: Callable,
   Returns:
     pytree with same structure as ``b``.
   """
-  def _matvec(x):
-    """Computes A^T A x."""
-    return _normal_matvec(matvec, x)
+  if init is None:
+    example_x = b  # This assumes that matvec is a square linear operator.
+  else:
+    example_x = init
+
+  try:
+    rmatvec = _make_rmatvec(matvec, example_x)
+    square = True
+  except TypeError:
+    square = False
+    raise TypeError("The initialization `init` of solve_normal_cg is "
+                    "compulsory when `matvec` is nonsquare. It should "
+                    "have the same pytree structure as a solution. "
+                    "Typically, a pytree filled with zeros should work.")
+
```

Comment:
It seems that the variable `square` is always `True` if we make it past this `try` block, even if we aren't in the square case. I think it still works out because of the previous comment about `_normal_matvec` being correct for non-square matvecs. Want to simplify or rework this function's logic? Is one way of doing `normal_matvec` preferred in the non-square case?

Reply:
Good catch, you're right `_normal_matvec` works in the nonsquare case too.